### PR TITLE
[21188] Example Refactor: Request-Reply 

### PR DESCRIPTION
**Pull request by EduPonz**
## Description

This PR refactors the request-reply example.
It also adds a `Add DomainParticipantFactory::get_participant_extended_qos_from_default_profile`

## Contributor Checklist

- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- [x] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- _N/A_: Any new/modified methods have been properly documented using Doxygen. 
- _N/A_: Any new configuration API has an equivalent XML API (with the corresponding XSD extension) 
- _N/A_: Changes are backport compatible: they do **NOT** break ABI nor change library core behavior. 
- _N/A_: Changes are API compatible. 
- [x] New feature has been added to the `versions.md` file (if applicable).
- _N/A_: New feature has been documented/Current behavior is correctly described in the documentation. 
    
- _N/A_: Applicable backports have been included in the description.

## Reviewer Checklist

- [x] The PR has a milestone assigned.
- [x] The title and description correctly express the PR's purpose.
- [x] Check contributor checklist is correct.
- _N/A_: If this is a critical bug fix, backports to the critical-only supported branches have been requested.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

File: examples/cpp/request_reply/CLIParser.hpp [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+     * @warning This method finishes the execution of the program if the input arguments are invalid
+     */
+    static std::int16_t consume_int16_argument(
+            const std::string& arg)
+    {
+        std::int16_t value = 0;
+
+        try
+        {
+            int input = std::stoi(arg);
```
Comment (elianalf) [inline examples/cpp/request_reply/CLIParser.hpp:247]: Referring to one of your suggestions [here](https://github.com/eProsima/Fast-DDS/pull/4999#discussion_r1654966934), wouldn't it be better to use a unsigned long.

File: examples/cpp/request_reply/CLIParser.hpp [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+
+            if (input < std::numeric_limits<std::int16_t>::min() ||
+                    input > std::numeric_limits<std::int16_t>::max())
+            {
+                throw std::out_of_range("int16 argument out of range");
+            }
+
+            value = static_cast<std::int16_t>(input);
```
Comment (elianalf) [inline examples/cpp/request_reply/CLIParser.hpp:255]: Why are we using a `int16_t` and not a type whose range is bigger?

**Review by elianalf — CHANGES_REQUESTED** [review]
Partial review.
The include of the subdirectory `RequestReplyExample` has not been removed from /exampels/cpp/dds/CMakeLists.txt.

File: examples/cpp/request_reply/CLIParser.hpp (hunk 1 above)
Comment (EduPonz) [reply examples/cpp/request_reply/CLIParser.hpp:247]: I don't think so. I'm just limitting the input numbers, which is documented in the help. Here, I'm assuming that `int` would either be `int16_t` (less probable) or `int32_t` (which is what it is in my implementation)

File: examples/cpp/request_reply/CLIParser.hpp (hunk 2 above)
Comment (EduPonz) [reply examples/cpp/request_reply/CLIParser.hpp:255]: I did not think it necessary

File: examples/cpp/request_reply/ServerApp.cpp [hunk 3]
```diff
@@ -0,0 +1,506 @@
...
+    if (nullptr == participant_)
+    {
+        throw std::runtime_error("Participant initialization failed");
+    }
+}
+
+template<>
+Topic* ServerApp::create_topic<CalculatorRequestTypePubSubType>(
+        const std::string& topic_name,
+        TypeSupport& type)
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply examples/cpp/request_reply/ServerApp.cpp:243]: I want class member function to not have it, as the class instance already does. The template that has is is just a free function in the cpp to group code.

File: examples/cpp/request_reply/ServerApp.cpp [hunk 4]
```diff
@@ -0,0 +1,506 @@
...
+                });
+
+        if (!is_stopped())
+        {
+            // Process one request per iteration
+            Request request = requests_.front();
+            requests_.pop();
+
+            rtps::GuidPrefix_t client_guid_prefix = rtps::iHandle2GUID(request.info.publication_handle).guidPrefix;
+            request_reply_debug("ServerApp", "Processing request from client " << client_guid_prefix);
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply examples/cpp/request_reply/ServerApp.cpp:352]: We decided to keep it, as in release it will not be there

**Comment by EduPonz** [comment]
All CI except the example's test passed for commit 05ba0c0.

I've increased the timeouts and ask for a manual run on the example's test [here](https://github.com/eProsima/Fast-DDS/actions/runs/9885796788)

**Comment by EduPonz** [comment]
Another manual CI [here](https://github.com/eProsima/Fast-DDS/actions/runs/9904977664)

File: test/examples/test_request_reply_isolated.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+        for line in out:
+            for client in expected_responses:
+                if client in line and 'Reply received from server' in line:
+                    responses[client] = parse_response(responses[client], line)
+
+        for client in responses:
+            if responses[client] != expected_responses[client]:
+                ret = False
+                print('ERROR: ' + client + ' expected "' + expected_responses[client] + '" but received "' + responses[client] + '"')
```
Comment (elianalf) [inline test/examples/test_request_reply_isolated.py:90]: It didn't print it 
```suggestion
                print('ERROR: ' + str(client) + ' expected "' + str(expected_responses[client]) + '" but received "' + str(responses[client]) + '"')
```

**Review by elianalf — CHANGES_REQUESTED** [review]

**Review by elianalf — APPROVED** [review]

**Pull request merged**